blockprod/src/detail/job_manager.rs

```diff
@@ -0,0 +1,358 @@
+                    log::info!("Error sending new job exists error: {e:?}");
+                }
+            } else {
+                jobs.insert(job_key.clone(), JobHandle { cancel_sender });
+
+                if let Err(e) = result_sender.send(Ok(job_key)) {
+                    log::info!("Error sending new job event: {e:?}");
+                }
+            }
+        }
+    }
+
+    pub async fn stop_job(&mut self, job_key: Option<JobKey>) -> Result<usize, JobManagerError> {
+        let (result_sender, result_receiver) = oneshot::channel();
+
+        if self.stop_job_sender.send((job_key, result_sender)).is_err() {
+            return Err(JobManagerError::FailedToSendStopJobEvent);
+        }
+
+        tokio::select! {
+            result = result_receiver => match result {
+                Ok(v) => Ok(v),
+                Err(_) => Err(JobManagerError::FailedToStopJobs),
+            }
+        }
+    }
+
+    fn stop_job_handler(
+        jobs: &mut BTreeMap<JobKey, JobHandle>,
+        event: Option<(Option<JobKey>, oneshot::Sender<usize>)>,
+    ) {
+        if let Some((job_key, result_sender)) = event {
+            let mut stop_jobs = Vec::new();
+
+            match job_key {
+                Some(job_key) => {
+                    if let Some(job_handle) = jobs.remove(&job_key) {
+                        stop_jobs.push((job_key, job_handle));
+                    }
+                }
+                None => {
+                    while let Some((job_key, job_handle)) = jobs.pop_first() {
+                        stop_jobs.push((job_key, job_handle));
+                    }
+
+                    log::info!("Cancelling {} jobs", stop_jobs.len());
+                }
+            }
+
+            let stop_count = stop_jobs.len();
+
+            for (job_key, job_handle) in stop_jobs.drain(..) {
+                let _ = job_handle.cancel_sender.send(());
+                log::info!("Stopped mining job for tip {}", job_key.current_tip_id());
+            }
+
+            if let Err(e) = result_sender.send(stop_count) {
+                log::info!("Error sending stop jobs count: {e:?}");
+            }
+        }
+    }
+
+    fn shutdown_handler(
+        jobs: &mut BTreeMap<JobKey, JobHandle>,
+        event: Option<oneshot::Sender<usize>>,
+    ) {
+        if let Some(result_sender) = event {
+            log::info!("Stopping job manager");
+            Self::stop_job_handler(jobs, Some((None, result_sender)));
+        }
+    }
+}
+
+impl Drop for JobManager {
+    fn drop(&mut self) {
+        let (result_sender, result_receiver) = oneshot::channel();
+
+        if let Err(e) = self.shutdown_sender.send(result_sender) {
+            log::info!("Error sending shutdown during job manager drop: {e:?}");
+            return;
+        }
+
+        tokio::spawn(async move {
+            tokio::select! {
+                _ = result_receiver => {},
+            }
+        });
```

Comment:
What's the purpose of this extra task? Would just closing the receiver work too?

Reply:
I'm guessing here... I think the purpose is to await on the result_receiver so that it eventually closes. The right way to do this, assuming having the task is OK, like I mentioned in another comment, is to await on result_receiver because it's a future, which is what `select!` will do. Besides that, the problem here is that fn drop is not async fn, which means you can't await. So Alfie found the only way to do this is by spawning an async task to do this, hoping that before the program shuts down, all tasks will be collected. Unlikely, but there's no better way I see.

To be fair, in a world where async makes sense, unlike what rust async has become, a blocking await would be the right way to go... but I think blocking-await can cause deadlocks... so here we are. I'm always afraid of suggesting blocking actions in async.

Reply:
Yep. As above, this wasn't as straight forward so had to jump through some hoops. Sam's doing some changes now, but if he hasn't worked out how to do RAII, I have a drop-esque work around that seems to work as I wanted.